File: test/python/hm_bethe.py
import numpy as np
import scipy as sp
from edipy import global_env as ed
import mpi4py
from mpi4py import MPI
import os,sys

#INIT MPI 
comm = MPI.COMM_WORLD
rank = comm.Get_rank()
print("I am process",rank,"of",comm.Get_size())
master = (rank==0)

def dens_bethe(x,d):
  root=(1-(x/d)**2)+0.j
  root=np.sqrt(root)
  dens_bethe=(2/(np.pi*d))*root
  return dens_bethe.real

#READ ED INPUT:
ed.read_input("inputED.conf")
if(ed.Nspin!=1 or ed.Norb!=1):
    print("This test code is for Nspin=1 + Norb=1.")
    ed.Nspin=1
    ed.Norb=1
Le      = 1000
wmixing = 0.3
wband   = 1.0

#BUILD Density of States:
Eband,de = np.linspace(-wband,wband,Le,retstep=True)
Dband = dens_bethe(Eband,wband)

#BUILD frequency array:
wm = np.pi/ed.beta*(2*np.arange(ed.Lmats)+1)
wr = np.linspace(ed.wini,ed.wfin,ed.Lreal)


#Allocate minimal required arrays:
#ALL functions must have shape [Nspin,Nspin,Norb,Norb(,L)]:
Smats=np.zeros((ed.Nspin,ed.Nspin,ed.Norb,ed.Norb,ed.Lmats),dtype='complex',order='F')
Gmats=np.zeros((ed.Nspin,ed.Nspin,ed.Norb,ed.Norb,ed.Lmats),dtype='complex',order='F')
Sreal=np.zeros((ed.Nspin,ed.Nspin,ed.Norb,ed.Norb,ed.Lreal),dtype='complex',order='F')
Greal=np.zeros((ed.Nspin,ed.Nspin,ed.Norb,ed.Norb,ed.Lreal),dtype='complex',order='F')
Delta=np.zeros((ed.Nspin,ed.Nspin,ed.Norb,ed.Norb,ed.Lmats),dtype='complex',order='F')
Hloc =np.zeros((ed.Nspin,ed.Nspin,ed.Norb,ed.Norb),dtype='float',order='F')



#SETUP SOLVER
Nb=ed.get_bath_dimension()
bath = np.zeros(Nb,dtype='float',order='F')
ed.init_solver(bath)
bath_prev = np.copy(bath)

#DMFT CYCLE
converged=False;iloop=0
while (not converged and iloop<ed.Nloop ):
    iloop=iloop+1
    print("DMFT-loop:",iloop,"/",ed.Nloop)

    #Solve the EFFECTIVE IMPURITY PROBLEM (first w/ a guess for the bath)
    ed.solve(bath,Hloc)
    

    #Get Self-energies
    ed.get_sigma_matsubara(Smats)
    ed.get_sigma_realaxis(Sreal)

    #Compute the local gf:
    for i in range(ed.Lmats):
        zeta             = wm[i]*1j - Smats[0,0,0,0,i]
        Gmats[0,0,0,0,i] = sum(Dband/(zeta - Eband))*de

    for i in range(ed.Lreal):
        zeta             = wr[i]+ed.eps*1j - Sreal[0,0,0,0,i]
        Greal[0,0,0,0,i] = sum(Dband/(zeta - Eband))*de
        
    if(rank==0):
        np.savetxt('Gloc_iw.dat', np.transpose([wm,Gmats[0,0,0,0,:].imag,Gmats[0,0,0,0,:].real]))
        np.savetxt('Gloc_realw.dat', np.transpose([wr,Greal[0,0,0,0,:].imag,Greal[0,0,0,0,:].real]))


    #Get the Delta function and FIT:    
    Delta[0,0,0,0,:] = 0.25*wband*Gmats[0,0,0,0,:]
    if(rank==0):
        np.savetxt('Delta_iw.dat', np.transpose([wm,Delta[0,0,0,0,:].imag,Delta[0,0,0,0,:].real]))
    ed.chi2_fitgf(Delta,bath,ispin=1,iorb=1)
    
    if(iloop>1):
        bath = wmixing*bath + (1.0-wmixing)*bath_prev
    bath_prev=np.copy(bath)
 
    err,converged=ed.check_convergence(Delta[0,0,0,0,:],ed.dmft_error,1,ed.Nloop)

print("Done...")

File: test/fortran/hm_bethe.f90
program hm_bethe
  USE EDIPACK
  USE SCIFOR
  USE MPI
  implicit none
  !INPUT:
  integer                                     :: Le
  real(8)                                     :: wmixing
  real(8)                                     :: Wband
  !
  !Bath:
  integer                                     :: Nb
  real(8),dimension(:),allocatable            :: Bath,Bath_prev
  !Local Variables:
  real(8),dimension(:,:,:,:),allocatable      :: Hloc
  real(8),dimension(:),allocatable            :: Ebands
  real(8),dimension(:),allocatable            :: Dbands

  real(8)                                     :: de
  !Local Functions:  
  complex(8),allocatable,dimension(:,:,:,:,:) :: Smats,Sreal
  complex(8),allocatable,dimension(:,:,:,:,:) :: Gmats,Greal
  complex(8),allocatable,dimension(:,:,:,:,:) :: Delta
  !Other:
  integer                                     :: iloop,iorb,i
  logical                                     :: converged
  complex(8)                                  :: zeta
  real(8),dimension(:),allocatable            :: wm,wr
  integer                                     :: comm,rank
  logical                                     :: master


  !MPI Wrapper from SciFortran:
  call init_MPI()
  comm = MPI_COMM_WORLD
  call StartMsg_MPI(comm)
  rank = get_Rank_MPI(comm)
  master = get_Master_MPI(comm)


  !READ ED INPUT:
  call ed_read_input('inputED.conf')
  if(Nspin/=1.OR.Norb/=1)stop "This test code is for Nspin=1 + Norb=1."
  Le     = 1000
  wmixing= 0.5d0
  Wband  = 1d0

  !BUILD Density of States:
  !linspace, dens_bethe, arange, pi in SciFortran
  allocate(Ebands(Le))
  allocate(Dbands(Le))
  Ebands = linspace(-Wband,Wband,Le,mesh=de)
  Dbands = dens_bethe(Ebands,Wband)*de

  !BUILD frequency array:
  allocate(wm(Lmats),wr(Lreal))
  wm = pi/beta*(2*arange(1,Lmats)-1)
  wr = linspace(wini,wfin,Lreal)


  !ALL functions must have shape [Nspin,Nspin,Norb,Norb(,L)]:
  allocate(Delta(Nspin,Nspin,Norb,Norb,Lmats))
  allocate(Smats(Nspin,Nspin,Norb,Norb,Lmats))
  allocate(Gmats(Nspin,Nspin,Norb,Norb,Lmats))
  allocate(Sreal(Nspin,Nspin,Norb,Norb,Lreal))
  allocate(Greal(Nspin,Nspin,Norb,Norb,Lreal))
  allocate(Hloc(Nspin,Nspin,Norb,Norb))
  Hloc = 0d0


  !SETUP SOLVER
  Nb=ed_get_bath_dimension()
  allocate(bath(Nb),bath_prev(Nb))
  call ed_init_solver(bath)

  !DMFT CYCLE
  iloop=0;converged=.false.
  do while(.not.converged.AND.iloop<nloop)
     iloop=iloop+1
     if(master)write(*,*)"DMFT-loop:",iloop,"/",nloop
     !
     !Solve the EFFECTIVE IMPURITY PROBLEM (first w/ a guess for the bath)
     call ed_solve(bath,Hloc)
     !
     !Get Self-energies
     call ed_get_sigma_matsubara(Smats)
     call ed_get_sigma_realaxis(Sreal)
     !
     !Compute the local gf:
     do i=1,Lmats
        zeta    = xi*wm(i) + xmu  - Smats(1,1,1,1,i)
        Gmats(1,1,1,1,i) = sum(Dbands(:)/(zeta - Ebands(:)))
     enddo
     do i=1,Lreal
        zeta     = cmplx(wr(i),eps) + xmu - Sreal(1,1,1,1,i)
        Greal(1,1,1,1,i) = sum(Dbands(:)/(zeta - Ebands(:)))
     enddo
     if(master)call splot("Gloc_iw.dat",wm,Gmats(1,1,1,1,:))    !splot is in SciFortran
     if(master)call splot("Gloc_realw.dat",wr,Greal(1,1,1,1,:)) !splot is in SciFortran
     !
     !Get the Delta function and FIT:
     Delta(1,1,1,1,:) = 0.25d0*Wband*Gmats(1,1,1,1,:)        
     call ed_chi2_fitgf(Delta,Bath,ispin=1)
     !
     !MIXING:
     if(iloop>1)Bath = wmixing*Bath + (1.d0-wmixing)*Bath_prev
     Bath_prev=Bath
     !
     !Check convergence
     converged = check_convergence(Delta(1,1,1,1,:),dmft_error,nsuccess,nloop)
     if(master)write(*,*)"---------"
  enddo

  call finalize_MPI()


contains

  function check_convergence(Xnew,eps,N1,N2) result(convergence)
    complex(8),intent(in)           :: Xnew(:)
    real(8),intent(in)              :: eps
    integer,intent(in)              :: N1,N2
    logical                         :: convergence  
    integer                         :: i,j,Msum
    real(8)                         :: err
    real(8)                         :: M,S
    complex(8),save,allocatable     :: Xold(:)
    integer,save                    :: success=0,check=1
    character(len=100)              :: file_
    file_='error.err'
    Msum=size(Xnew)
    if(.not.allocated(Xold))then
       allocate(Xold(Msum))
       Xold=0.d0
    endif
    S=0.d0 ; M=0.d0
    do i=1,Msum
       M=M + abs(Xnew(i)-Xold(i))
       S=S + abs(Xnew(i))
    enddo
    err= M/S
    Xold=Xnew
    if(master)then
       open(10,file=reg(file_),position="append")
       write(10,"(I5,ES15.7)")check,err
       close(10)
    endif
    if(err < eps)then
       success=success+1
    else
       success=0
    endif
    convergence=.false.
    if(success > N1)convergence=.true.
    if(check>=N2)then
       if(master)then
          open(10,file="ERROR.README")
          write(10,*)""
          close(10)
          write(*,"(A,I4,A)")"Not converged after",N2," iterations."
       endif
    endif
    if(convergence)then
       if(master)write(*,"(A,ES15.7)")bold_green("error="),err
    else
       if(err < eps)then
          if(master)write(*,"(A,ES15.7)")bold_yellow("error="),err
       else
          if(master)write(*,"(A,ES15.7)")bold_red("error="),err
       endif
    endif
    check=check+1
  end function check_convergence


end program hm_bethe












